content/blog/2022-05-05-e2e-cv-pt3.md

```diff
@@ -0,0 +1,289 @@
+              run: |
+                cml runner \
+                    --cloud=aws \
+                    --cloud-region=us-east-1 \
+                    --cloud-type=g4dn.xlarge \
+                    --labels=cml-runner
+      ...
+      ```
+
+   2. deploy our experiment branch to a docker container on this machine:
+
+      ```yaml
+      ---
+      train-model:
+        needs: deploy-runner
+        runs-on: [self-hosted, cml-runner]
+        container:
+          image: iterativeai/cml:0-dvc2-base1
+          options: --gpus all
+        environment: cloud
+        permissions:
+          contents: read
+          id-token: write
+        steps:
+          - uses: actions/checkout@v2
+          - uses: actions/setup-python@v2
+            with:
+              python-version: '3.9'
+      ```
+
+   3. rerun the entire DVC pipeline & push metrics back to GitHub;
+
+      ```yaml
+      ...
+      - name: dvc-repro-cml
+              env:
+                REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+              run: |
+                # Install dependencies
+                pipenv install --skip-lock
+                pipenv run dvc pull
+                pipenv run dvc exp run
+                pipenv run dvc push
+      ...
+      ```
+
+   4. open a pull request and post a report to it that contains a table with
+      metrics and model outputs (segmentation masks) on a few test images:
+
+      ```bash
+      ...
+      # Open a pull request
+      cml pr dvc.lock metrics.json training_metrics.json training_metrics_dvc_plots/**
+      # Create CML report
+      echo "## Metrics" >> report.md
+      pipenv run dvc metrics show --md >> report.md
+      echo "## A few random test images" >> report.md
+      for file in $(ls data/test_preds/ |sort -R |tail -20); do
+        cml publish data/test_preds/$file --md >> report.md
+      done
+      cml send-comment --pr --update report.md
+      ...
+      ```
+
+      The report structure is fully customizable. Below is an example of what
+      the PR and the CML report would look like in this case:
+
+      ![PR and CML report](/uploads/images/2022-05-05/pr_cml_report.png '=800')
+
+   At this point, we can assess the results in Iterative Studio and GitHub and
+   decide whether we want to accept the PR or keep experimenting.
+
+2. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/2-develop.yaml)
+
+   ![Workflow for deploying to the development environment](/uploads/images/2022-05-05/workflow_dev.png '=800')
+
+   Once we are happy with our model's performance on the experiment branch, we
+   can merge it into the development branch. This would trigger a different
+   CI/CD job that will:
+
+   1. retrain the model with the new parameters (as the `dev` branch might
+      contain changes not present in the `exp` branch). This step looks almost
+      identical to the third step in the above workflow.
```

Comment:
Maybe mention that if nothing changed, the model won't be retrained because of how dvc pipelines work.